### PR TITLE
Fix: Do not send vehicles towards incomplete PF nodes

### PR DESCRIPTION
**Pull request by vituscze**
## Motivation / Problem

When a path cannot be found, the pathfinder attempts to send the vehicle as close to its destination as possible. However, it is possible to set up a situation where the pathfinder ends up sending the vehicle *away* from the destination.

Here's a savegame which demonstrates this issue:  [overwrite_bug.zip](https://github.com/OpenTTD/OpenTTD/files/6496806/overwrite_bug.zip). If you start the vehicle, it chooses the track going *away* from its destination. If you remove the waypoint or the signal after it, the train chooses the correct track.

## Description

When the pathfinder is exploring a new graph node, it requests a fresh node from the underlying `NodeList` which can either produce a completely new node or reuse the last node (if that node was never added to the open list or was not marked as the destination).

It is possible to set up a situation where the `NodeList` reuses a node that's currently marked as the best intermediate node (`m_pBestIntermediateNode`). If the best node is then never updated, it will end up with whatever random data the next round of the pathfinder fills it with, which might be completely unrelated to the actual best node.

The attached savegame works by exploiting very long track segments (`ESRB_SEGMENT_TOO_LONG`). In particular, the are two ways of getting to the waypoint (key: `m_tile=3055 m_td=TRACKDIR_X_NE`) that end up having different `m_last_tile`s of their segments. When the pathfinder gets to the waypoint for the second time, it creates a new node, it sees that its last tile is closer to the destination and sets it as the best intermediate node. Then it finds out that the key is already present in the closed list and the new node is never added to the open list (and thus finalized).

The fix to this issue is simple: the best intermediate node is only updated once we're sure that the construction of the node is finalized (i.e. it is inserted into the open list).

## Limitations

A case could be made for changing the best intermediate node to the `openNode` or `closedNode`, but the content of the open/closed node couldn't be changed to match the newly created node (in particular the distance to the destination would be wrong). This is very unlikely to matter, but I'm open to suggestions.

Another possibility is to use `ItemList`'s `FoundBestNode` to ensure the node isn't reused, but that seems like a misuse of that function.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/pathfinder/yapf/yapf_base.hpp [hunk 1]
```diff
@@ -249,10 +249,6 @@ class CYapfBaseT {
 			return;
 		}
 
-		if (m_max_search_nodes > 0 && (m_pBestIntermediateNode == nullptr || (m_pBestIntermediateNode->GetCostEstimate() - m_pBestIntermediateNode->GetCost()) > (n.GetCostEstimate() - n.GetCost()))) {
-			m_pBestIntermediateNode = &n;
-		}
-
```
Comment (rubidium42) [inline src/pathfinder/yapf/yapf_base.hpp:255]: Shouldn't this code, though with `*openNode` instead of `n`, also be used in the branch below where it calls `InsertOpenNode` (line 266/262)? It feels like `*openNode` could then be the best intermediate node in that branch. Granted, doing it on `n` is definitely wrong here and there.
Or in other words, shouldn't this "is this the best intermediate node logic" be called just after both calls to `InsertOpenNode`?

Comment (vituscze) [reply src/pathfinder/yapf/yapf_base.hpp:255]: Yeah, this is one possibility. My main issue with this was that if you reach a new best intermediate node and find out its key is already in the open or closed list, you can only "fix" this in the case of an existing worse node in the open list. So I opted for setting the best intermediate node only when the node is truly fresh, in order to avoid surprises.

But if you think the best intermediate node should be set in the case of a worse open node, I can adjust.

Comment (vituscze) [reply src/pathfinder/yapf/yapf_base.hpp:255]: I suppose this really just depends on whether the best intermediate node is supposed to consider any node or only nodes that end up in the open/closed list.

Comment (rubidium42) [reply src/pathfinder/yapf/yapf_base.hpp:255]: My comment is mostly based on the "i.e. it is inserted into the open list" remark in your initial description, as I saw another place the InsertOpenNode got called and the reconsidering did not happen.
Sadly enough I could not trigger that branch of the code with a simple game, so I have no way to know whether it not getting updated in the branch would constitute to a regression in existing behavior or not. And thus whether not doing it might be surprising behavior or not.

In conclusion: I do agree that the current location where the best intermediate node is considered is definitely wrong, but I'm not sure whether only doing it in one of the InsertOpenNode cases causes unwanted surprises.

Comment (vituscze) [reply src/pathfinder/yapf/yapf_base.hpp:255]: I think the only way you can get to this branch of the code is with smaller initial segments that have uneven amount of signals which are then followed by a long, uninterrupted segments that end due to ESRB_SEGMENT_TOO_LONG. I've managed to trigger a breakpoint in this part of code when I was testing something unrelated on the savegame from #9271 (happens to trains around 931x702), but it took me quite a while to actually construct a situation where this changes the path vehicles take.

In pretty much all non-artificial situations I've tried, the pathfinder simply extends the best ESRB_SEGMENT_TOO_LONG node, sets a new best intermediate node and the problem disappears (since ESRB_SEGMENT_TOO_LONG nodes generally pass enough signals to get past the lookahead).

---

I suppose saying that the behavior would be surprising might be overstating it somewhat. Basically, when I opened the PR I was thinking about `m_pBestIntermediateNode` in the same way as `m_pBestDestNode`, i.e. it should be any node closest to the destination, even if it never ends up in the A* graph.

The best destination node can also end up in a similar situation (i.e. never being added to the open list), but it notifies the node list not to reuse it. While this could also be used to fix this problem, it seems extremely wasteful as most intermediate nodes are then thrown away. It would also result in some other annoyances with, for example, branch pruning.

So, we're left with the possibility of setting the best intermediate node in 2 out of the 4 cases that can occur (fresh node, worse open, better open, closed) and anything we do here is going to be slightly wrong and thus, in some sense, surprising. Handling only the fresh node seemed a bit less wrong than the other options since it's somewhat analogous to the pathfinder always picking the first closest node as the intermediate, regardless of cost.

On the other hand, if `m_pBestIntermediateNode` is only supposed to contain nodes that actually end up in the A* graph, then only handling the fresh node and worse open node would be correct. After thinking about it some more, I'm actually liking this second option a bit better, but I also haven't been around when YAPF was added so perhaps that interpretation is wrong.

Comment (vituscze) [reply src/pathfinder/yapf/yapf_base.hpp:255]: I'm now convinced that doing the latter (i.e. only consider nodes that end up in the A* graph) is the better option.

**Review by rubidium42 — APPROVED** [review]
